### PR TITLE
Added case on 'terminated-but-not-yet-deleted' for Admit.

### LINKED ISSUE #47867: Scheduler and kubelet should consider resources of terminated-but-not-yet-deleted pods to still be in use.
@vishh pointed out that for scheduling and kubelet admission, we should consider resources of terminated-but-not-yet-deleted pods to still be in use, because it may take an arbitrary amount of time for the system to deallocate resources. We didn't check to see if this is currently how it is implemented or not.

cc/ @vishh @bsalamat 

/sig scheduling

### PR DESCRIPTION
**Pull request by k82cn**
**What this PR does / why we need it**:
Added case on 'terminated-but-not-yet-deleted' for Admit.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #47867 

**Release note**:

```release-note-none
```

### CODE REVIEW

**Comment by k82cn** [comment]
/unassign @sjpotter @feiskyer 

Will re-assign to you after adding UT case.

**Comment by k82cn** [comment]
cc/ @vishh @bsalamat @davidopp

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -743,6 +743,20 @@ func (kl *Kubelet) podIsTerminated(pod *v1.Pod) bool {
 	return status.Phase == v1.PodFailed || status.Phase == v1.PodSucceeded || (pod.DeletionTimestamp != nil && notRunning(status.ContainerStatuses))
 }
 
+// podIsFinished returns true if pod is in the finished state ("Failed" or "Succeeded").
+func (kl *Kubelet) podIsFinished(pod *v1.Pod) bool {
```
Comment (bsalamat) [inline pkg/kubelet/kubelet_pods.go]: nit: I am not sure about `podIsFinished`. It is kind of confusing that we have both "podIsTerminated" and "podIsFinished". I don't have any better suggestion though.

File: pkg/kubelet/kubelet_pods.go [hunk 2]
```diff
@@ -729,7 +729,7 @@ func (kl *Kubelet) getPullSecretsForPod(pod *v1.Pod) []v1.Secret {
 	return pullSecrets
 }
 
-// Returns true if pod is in the terminated state ("Failed" or "Succeeded").
+// podIsTerminated returns true if pod is in the terminated state ("Failed", "Succeeded" or "DeletionTimestamp" not nil).
```
Comment (bsalamat) [inline pkg/kubelet/kubelet_pods.go]: The new comment is not accurate. This method return true when "DeletionTimestamp" is not nil AND pod containers are not running.

File: pkg/kubelet/kubelet_pods.go [hunk 3]
```diff
@@ -743,6 +743,20 @@ func (kl *Kubelet) podIsTerminated(pod *v1.Pod) bool {
 	return status.Phase == v1.PodFailed || status.Phase == v1.PodSucceeded || (pod.DeletionTimestamp != nil && notRunning(status.ContainerStatuses))
 }
 
+// podIsFinished returns true if pod is in the finished state ("Failed" or "Succeeded").
+func (kl *Kubelet) podIsFinished(pod *v1.Pod) bool {
+	var status v1.PodStatus
```
Comment (bsalamat) [inline pkg/kubelet/kubelet_pods.go]: Do you need to declare this variable explicitly?

**Review by bsalamat — COMMENTED** [review]
My main concern is that I think the existing logic of Kubelet is correct. With your change, Kubelet would consider the terminated-but-not-yet-deleted pods as active and would account their resources when admitting a new pod. IMO we shouldn't account resources of terminated pods.

**Comment by k82cn** [comment]
@bsalamat , we should account `terminated-but-not-yet-deleted` for Admit, similar to scheduler. hm.. I think terminating is more accurate. But in kubelet, it has already considered this case by `pod.DeletionTimestamp != nil && notRunning(status.ContainerStatuses)`.

I updated comments to the issue, I'll close this PR if no other comments.

**Comment by bsalamat** [comment]
> we should account terminated-but-not-yet-deleted for Admit, similar to scheduler. 

I am not sure if this is the right approach. IMO, kubelet admission logic will become unnecessarily more restrictive after this PR.
@davidopp What do you think?

**Comment by k82cn** [comment]
> IMO, kubelet admission logic will become unnecessarily more restrictive after this PR.

do you mean the latest version?

**Comment by davidopp** [comment]
The current version of this PR doesn't seem to change the kubelet admission logic -- maybe a commit is missing?

**Comment by k82cn** [comment]
I think Bobby is talking about previous version :).

**Comment by bsalamat** [comment]
Yes, I was talking about the previous version of this PR. This one only improves a test.

/lgtm

**Comment by k82cn** [comment]
@timstclair , would you help to approve that?

**Comment by k82cn** [comment]
@timstclair , more comments?

**Comment by k82cn** [comment]
@tallclair , I think you need to update OWNER in kubelet :).

**Pull request merged**